lib/firebase_functions_interop.dart

```diff
@@ -456,3 +460,136 @@ class PubsubEvent extends Event<Message> {
           timestamp: timestamp,
         );
 }
+
+class StorageFunctions {
+  const StorageFunctions._();
+
+  /// Registers a Cloud Function scoped to a specific storage [bucket].
+  BucketBuilder bucket(String path) =>
+      new BucketBuilder._(_js.storage.bucket(path));
+
+  /// Registers a Cloud Function scoped to the default storage bucket for the project.
+  ObjectBuilder object() => new ObjectBuilder._(_js.storage.object());
+}
+
+class BucketBuilder {
+  @protected
+  final js.BucketBuilder nativeInstance;
+
+  BucketBuilder._(this.nativeInstance);
+
+  /// Storage object builder interface scoped to the specified storage bucket.
+  ObjectBuilder object() => new ObjectBuilder._(nativeInstance.object());
+}
+
+class ObjectBuilder {
+  @protected
+  final js.ObjectBuilder nativeInstance;
+
+  ObjectBuilder._(this.nativeInstance);
+
+  /// Event handler which fires every time a Google Cloud Storage change occurs.
+  js.CloudFunction onChange(FutureOr<void> handler(StorageEvent event)) {
+    dynamic wrapper(js.Event jsEvent) => _handleEvent(jsEvent, handler);
+    return nativeInstance.onChange(allowInterop(wrapper));
+  }
+
+  dynamic _handleEvent(
+      js.Event jsEvent, FutureOr<void> handler(StorageEvent event)) {
+    final StorageEvent event = new StorageEvent(
+      data: new ObjectMetadata(jsEvent.data),
+      eventId: jsEvent.eventId,
+      eventType: jsEvent.eventType,
+      params: dartify(jsEvent.params),
+      resource: jsEvent.resource,
+      timestamp: DateTime.parse(jsEvent.timestamp),
+    );
+    var result = handler(event);
+    if (result is Future) {
+      return futureToPromise(result);
+    }
+    return null;
+  }
+}
+
+class ObjectMetadata {
+  ObjectMetadata(js.ObjectMetadata this.nativeInstance);
+
+  @protected
+  final js.ObjectMetadata nativeInstance;
+
+  String get bucket => nativeInstance.bucket;
+
+  String get cacheControl => nativeInstance.cacheControl;
+
+  int get componentCount => nativeInstance.componentCount;
+
+  String get contentDisposition => nativeInstance.contentDisposition;
+
+  String get contentEncoding => nativeInstance.contentEncoding;
+
+  String get contentLanguage => nativeInstance.contentLanguage;
+
+  String get contentType => nativeInstance.contentType;
+
+  js.CustomerEncryption get customerEncryption {
+    final dartified = dartify(nativeInstance.customerEncryption);
+    return new js.CustomerEncryption(
+        encryptionAlgorithm: dartified['encryptionAlgorithm'],
+        keySha256: dartified['keySha256']);
+  }
```

Comment:
I like this version too. Two things though:

1. `dartify` may return `null` (because this field is nullable) so I'd add `if (dartified == null) return null;`
2. The `CustomerEncryption` class is not a pure JS object the way it's declared right now, so we either need to update it to be pure JS or make it pure Dart.

I understand JS interop can be confusing, it took me a few months to get a good grasp on all the basics. Here is some background:

JS interop allows defining interface for a pure JS object with a static/strict shape. For this it: (1) must be declared `@anonymous`, (2) have all fields `external` and (3) have `external factory` constructor with all parameters named. The (3) is optional and only needed if we want to create a new instance of this object in Dart code.

Example:

```dart
@JS()
@anonymous
abstract class Stuff {
  external String get thing;
  external factory Stuff({String thing});
}
```

What this does when compiled to JS is if we have for instance: 

```dart
var stuff = new Stuff(thing: "something");
print(stuff.thing);
```
 it gets compiled to 

```js
var stuff = {"thing": "something"};
console.log(stuff.thing);
``` 

(roughly speaking).

However if constructor is not declared as `external` Dart will try to create it's own version of this class which includes some extra boilerplate for identity management, `toString` and such.

So we have two options in this case, as I see it:

1. Declare everything in `CustomerEncryption` external and simply proxy the value from nativeInstance.
2. Remove `@JS()` and `@anonymous` from `CustomerEncryption`, move it to the main library file, revert update to exports (as it will be exported automatically), check for `null` in the `dartified` var.

I originally thought to go with (1) but don't have strong preference so I'd let you make the call.